sw/device/tests/sensor_ctrl_alerts.c

```diff
@@ -0,0 +1,229 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_alert_handler.h"
+#include "sw/device/lib/dif/dif_flash_ctrl.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_sensor_ctrl.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/flash_ctrl_testutils.h"
+#include "sw/device/lib/testing/pwrmgr_testutils.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sensor_ctrl_regs.h"  // Generated.
+
+/**
+ * This test checks that incoming ast events can be
+ * configured as both recoverable and fatal.
+ * Further, this test checks that recoveralbe and fatal
+ * events are able to reach their proper alert_handler
+ * destination.
+ *
+ * Since fatal events do not stop firing once asserted,
+ * this test performs a self reset after every fatal
+ * event.  In order to keep track of how far the test
+ * has advanced, a non-volatile counter in flash is
+ * used to track current progress.
+ */
+static dif_rstmgr_t rstmgr;
+static dif_flash_ctrl_state_t flash_ctrl;
+static dif_sensor_ctrl_t sensor_ctrl;
+static dif_alert_handler_t alert_handler;
+
+/**
+ * The events_vector stores the tested events in bit strike fashion.
+ * This means at first the flash word is 0xFFFF_FFFF.
+ * As events are tested, they are marked as 0 in the vector,
+ * so we go from 0xFFFF_FFFF -> 0xFFFF_FFFE -> 0xFFFF_FFFC
+ */
+__attribute__((section(".non_volatile_scratch")))
+const volatile uint32_t events_vector = -1;
+
+/**
+ *  Extracts current event id from the bit-strike counter.
+ */
+static uint32_t event_to_test() {
+  uint32_t addr = (uint32_t)(&events_vector);
+  uint32_t val = abs_mmio_read32(addr);
+
+  for (size_t i = 0; i < 32; ++i) {
+    if (val >> i & 0x1) {
+      return i;
+    }
+  }
+  return -1;
+};
+
+/**
+ *  Increment flash bit-strike counter.
+ */
+static bool incr_flash_cnt(uint32_t tested_idx) {
+  uint32_t addr = (uint32_t)(&events_vector);
+
+  // set the tested bit to 0
+  uint32_t val = abs_mmio_read32(addr) & ~(1 << tested_idx);
```

Comment:
ah should i have done this a different way?

Reply:
> ah should i have done this a different way?

I'm not sure. It's an unusual case for which I don't have a feel for what is best for readability. It seems somewhat roundabout to go from a `const volatile uint32_t` to a `uint32_t` then back to a `volatile uint32_t*` `in abs_mmio_read32()`. But since you needed `addr` to be a `uint32_t` for `flash_ctrl_testutils_write()` anyway, I think I'll balk from worrying about the style, haha.

Reply:
sounds good to me :)